test/cpp/qps/client_async.cc

```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
       return nullptr;
   }
 }
+
 std::unique_ptr<Client> CreateGenericAsyncStreamingClient(
     const ClientConfig& args) {
   return std::unique_ptr<Client>(new GenericAsyncStreamingClient(args));
 }
 
+struct CallbackClientRpcData {
+  CallbackClientRpcData()
+      : thread_mu_(), thread_cv_(), rpc_done_(false), alarm_(nullptr) {}
+
+  ~CallbackClientRpcData() { delete alarm_; }
+
+  std::mutex thread_mu_;
+  std::condition_variable thread_cv_;
+  bool rpc_done_;
+  Alarm* alarm_;
+};
+
+class CallbackClient
+    : public ClientImpl<BenchmarkService::Stub, SimpleRequest> {
+ public:
+  CallbackClient(const ClientConfig& config)
+      : ClientImpl<BenchmarkService::Stub, SimpleRequest>(
+            config, BenchmarkStubCreator) {
+    num_threads_ = NumThreads(config);
+    ctxs_.resize(num_threads_);
+    threads_done_ = 0;
+    SetupLoadTest(config, num_threads_);
+  }
+
+  virtual ~CallbackClient() {}
+
+  virtual bool InitThreadFuncImpl(size_t thread_idx) { return true; }
+
+  virtual bool ThreadFuncImpl(Thread* t, size_t thread_idx) = 0;
+
+ protected:
+  size_t num_threads_;
+  std::mutex mu_;
+  std::condition_variable cv_;
+  std::vector<CallbackClientRpcData*> ctxs_;
+  size_t threads_done_;
+
+  void NotifyThread(size_t thread_idx) {
+    std::mutex* thread_mu = &ctxs_[thread_idx]->thread_mu_;
+    std::condition_variable* thread_cv = &ctxs_[thread_idx]->thread_cv_;
+    std::lock_guard<std::mutex> thread_lock(*thread_mu);
+    ctxs_[thread_idx]->rpc_done_ = true;
+    thread_cv->notify_one();
```

Comment:
Likewise why not directly use the original `cv` struct from the `ctx` rather than going through an additional pointer